### PR TITLE
Add pytorch lightning baseline

### PR DESCRIPTION
**Pull request by ojqbo**
added:
- PytorchLightning Datamodule built on top of Torch Dataset
  - training set is split into new training set and a validation set. The split is stratified based on patients that have cancer.
  - no test dataloader is provided
  - optional oversampling of cancer cases
- PytorchLightning Classifier with pF1 as Loss. Temporarily, a dummy model is used as a backbone.

### CODE REVIEW

**Comment by ojqbo** [comment]
dependency-review fails due to CVE:
pytorch-lightning1.8.6 (latest) requires tensorboardX = ">=2.2"
tensorboardx2.5.1 (latest) requires protobuf = ">=3.8.0,<=3.20.1"
protobuf has a CVE that was patched in version 3.20.2

tensorboardx pull request to update protobuf: https://github.com/lanpa/tensorboardX/pull/675

File: src/nn/data.py [hunk 1]
```diff
@@ -77,22 +79,114 @@ def __init__(self, labels_csv_path: Path, images_folder: Path) -> None:
...
+            patients_to_oversample = train_patients[
+                train_patients.cancer == 1
+            ].index.to_numpy()
+            oversample_multiplier = (
+                len(train_patients) // len(patients_to_oversample) - 1
+            )  # will raise value error if there are no patients with cancer
+            train_idxs_to_oversample = full_train_dataset.metadata[
+                full_train_dataset.metadata.patient_id.isin(patients_to_oversample)
+            ].index.to_numpy()
+            train_idxs = np.concatenate(
```
Comment (piotr-rarus) [inline src/nn/data.py:181]: Tutaj użyłbym `RandomOverSampler` z `imblearn`.

File: src/nn/data.py [hunk 2]
```diff
@@ -77,22 +79,114 @@ def __init__(self, labels_csv_path: Path, images_folder: Path) -> None:
...
+                [train_idxs] + [train_idxs_to_oversample] * oversample_multiplier
+            )
+        np.random.default_rng(seed=random_state).shuffle(train_idxs)
```
Comment (piotr-rarus) [inline src/nn/data.py:184]: Splitter powinien robić shuffle, zamiast tak z palca.

Comment (piotr-rarus) [inline src/nn/data.py:184]: Ziarna bym chyba wywalił i wyżej wołał `seed_everything`. Nie wiem, czy chcemy trzymać osobne ziarna.

File: src/nn/data.py [hunk 3]
```diff
@@ -77,22 +79,114 @@ def __init__(self, labels_csv_path: Path, images_folder: Path) -> None:
...
+        full_train_dataset = RSNABreastCancerTrainDataset(
+            images_dir=images_dir,
+            metadata_csv_path=metadata_csv_path,
+        )
+
+        # we can't just split indexes, we should split by patients:
+        patients_to_stratify = full_train_dataset.metadata.groupby(by="patient_id")[
+            ["cancer"]
+        ].max()  # a dataframe: index: "patient_id"; columns: ["cancer"]
+        skf = StratifiedKFold(n_splits=val_dataset_factor)
```
Comment (piotr-rarus) [inline src/nn/data.py:158]: Ostatecznie nie używamy na końcu kilku splitów, więc zrobiłbym zwykły podział ze stratyfikacją.

File: src/nn/lightning_model.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+        self.log(f"{prefix}/loss", loss)
+        for metric_name, metric in self.metrics.items():
+            metric_value = metric.compute()
+            metric.reset()
+            self.log(f"{prefix}/{metric_name}", metric_value)
+            if metric_name == "accuracy":
+                self.log(f"{prefix}/error", 1 - metric_value)
+
+
+class LightningClassifier(LightningBaseModule):
```
Comment (piotr-rarus) [inline src/nn/lightning_model.py:104]: Po co nam klasa bazowa? Hiperparametry możemy przepchać w konstruktorze, zamiast robić nowe klasy.

File: src/nn/lightning_model.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+
+    def _shared_eval(
+        self, batch: torch.Tensor, batch_idx: int, prefix: str
+    ) -> dict[str, torch.Tensor]:
+        x, y = batch
+        logits = self.backbone(x)
+        preds = logits.softmax(dim=1)
+        loss = self.supervised_criterion(preds, y)
+        for metric_name, metric in self.metrics.items():
+            metric.update(preds, y)
```
[the remark this answers is not in the document]
Comment (ojqbo) [reply src/nn/lightning_model.py:80]: train i val dzieją się osobno, po epoce trenującej nastąpi reset metryki, potem epoka val, obliczanie metryki, i jej reset

File: src/nn/lightning_model.py (hunk 4 above)
Comment (ojqbo) [reply src/nn/lightning_model.py:103]: Możemy zrobić jedną klasę. Miałem akurat taką templatkę, w której z klasy bazowej robiłem klasyfikator i coś na kształt GAN-a

File: src/nn/lightning_model.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (piotr-rarus) [reply src/nn/lightning_model.py:80]: Reset robi się automatycznie na epoch end. Czy do tego czasu nie będą się razem mielić batche train i val? Muszę sprawdzić w debuggerze.

[the remark this answers is not in the document]
Comment (piotr-rarus) [reply src/nn/lightning_model.py:80]: Tak, najpierw lecą wszystkie training stepy, później val stepy i dopiero jest epoch_end i reset. Musimy osobne metryki porobić.

File: src/nn/data.py (hunk 2 above)
Comment (ojqbo) [reply src/nn/data.py:184]: Obecna forma ma taką zaletę, że jeśli chcemy zbadać skuteczność architektury, to możemy odpalić kod na seed_everything i wagi będą miały inny punkt początkowy za każdym razem, przy czym podział danych będzie nietknięty. Jeśli seed_everything będzie dotyczyć też danych, nie będziemy w stanie przeprowadzić rzetelnego eksperymentu badającego samą architekturę. Inną sprawą jest, że i tak nie będziemy mieli czasu na taki eksperyment

Comment (ojqbo) [reply src/nn/data.py:184]: Oversampling robiłem z palca więc shufflowanie zbioru treningowego też, można to zmienić. Dataloader później też będzie miał opcję by zrobić reshuffle przed każdą epoką. `RandomOverSampler` z `imblearn` o którym wspomniałeś być może też robi to pod spodem. Celem shufflowania w tym miejscu było wymieszanie etykiet tak, by model podczas treningu nie napotkał długich sekwencji takich samych etykiet. Pomiędzy epokami dane nie muszą być reshufflowane i wg mnie jest to dobra praktyka - stąd wolałem reshufflować indeksy raz, i nie ustawiać już potem shufflowania w Dataloaderze

File: src/nn/classifier.py [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+    def init_metrics(self, num_classes: int) -> None:
+        self.metrics = {
+            "train": self.__get_split_metrics(num_classes=num_classes),
+            "val": self.__get_split_metrics(num_classes=num_classes),
+            "test": self.__get_split_metrics(num_classes=num_classes),
+        }
+
+        for split_name in self.metrics:
+            for metric_name, metric in self.metrics[split_name].items():
+                self.add_module(f"metrics/{split_name}/{metric_name}", metric)
```
Comment (piotr-rarus) [inline src/nn/classifier.py:37]: Czemu dodajemy metryki jako moduły?

File: src/nn/classifier.py [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+            ),
+            "recall": torchmetrics.Recall(
+                task="multiclass", num_classes=num_classes, average="macro"
+            ),
+            "f1-score": torchmetrics.F1Score(
+                task="multiclass", num_classes=num_classes, average="macro"
+            ),
+        }
+
+        if num_classes > 5:
```
Comment (piotr-rarus) [inline src/nn/classifier.py:55]: Będziemy mieli więcej klas, czy tylko binary zdrowy/rak ?

File: src/nn/classifier.py [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+    def configure_optimizers(self) -> dict[str, Any]:
+        optimizer = torch.optim.SGD(self.backbone.parameters(), lr=self.lr)
+        scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(
+            optimizer, mode="max", patience=5
+        )
+        return {
+            "optimizer": optimizer,
+            "lr_scheduler": {
+                "scheduler": scheduler,
+                "monitor": "val/accuracy",
```
Comment (piotr-rarus) [inline src/nn/classifier.py:71]: Czy nie powinniśmy raczej monitorować lossa?

File: src/nn/classifier.py (hunk 6 above)
Comment (ojqbo) [reply src/nn/classifier.py:37]: Żeby Lightning zajmował się obsługą zmiennych wewnętrznych metryki w zaeżności od urządzenia na którym uczymy (cpu/gpu). "Modular metrics are automatically placed on the correct device when properly defined inside a LightningModule"

File: src/nn/classifier.py (hunk 7 above)
Comment (ojqbo) [reply src/nn/classifier.py:55]: ciężko powiedzieć, ale ten fragment rzeczywiście można usunąć. Gdybyśmy mieli więcej klas to i tak musiało by być to w formie multi-class.

File: src/nn/classifier.py (hunk 8 above)
Comment (ojqbo) [reply src/nn/classifier.py:71]: Zdecydowanie tak, dobrze że zauważyłeś.

**Review by piotr-rarus — APPROVED** [review]
